### PR TITLE
JSAB2-89 Forgotten password

### PR DESCRIPTION
**Pull request by HanqingZ**
- add a forgetten password page
- User can enter whether username or email to find back their password

### CODE REVIEW

**Comment by HanqingZ** [comment]
Please use the branch JSAB2-89-forgotten-page-test to test
This branch create a button link to the forgotten password page

File: frontend/src/Login/actionCreator.js [hunk 1]
```diff
@@ -1,7 +1,22 @@
...
 export function login() {
   return {
     type: LOGIN,
   };
 }
+
+export function forgotPassword(input) {
+  if (input.includes('@')) {
+    return {
+      type: FORGOTPASSWORD,
```
Comment (kyya) [inline frontend/src/Login/actionCreator.js]: please add underscore between FORGOT and PASSWORD.

File: frontend/src/Login/authReducer.js [hunk 2]
```diff
@@ -8,6 +8,8 @@ export default function auth(state = initialState, action) {
   switch (action.type) {
     case LOGIN:
       return { ...state, isLoggedIn: true };
+    case FORGOTPASSWORD:
```
Comment (kyya) [inline frontend/src/Login/authReducer.js]: please add underscore between FORGOT and PASSWORD.

**Review by kyya — CHANGES_REQUESTED** [review]

**Review by GuJialu — APPROVED** [review]

File: frontend/src/Login/actionCreator.js [hunk 3]
```diff
@@ -1,7 +1,22 @@
...
+export const FORGOT_PASSWORD = 'forgotPassword';
 
 export function login() {
   return {
     type: LOGIN,
   };
 }
+
+export function forgotPassword(input) {
+  if (input.includes('@')) {
```
Comment (kyya) [inline frontend/src/Login/actionCreator.js]: If I have an email called '@@@@@@@@@', then what will happen?

File: frontend/src/Login/ForgottenPassword.jsx [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    <View style={styles.container}>
+      <ImageBackground
+        style={styles.background}
+        resizeMode="cover"
+        source={background}
+      >
+        <Text style={styles.header}>Forget Password</Text>
+        <View style={{ width: 300 }}>
+          <InputField
+            placeholder="input"
```
Comment (kyya) [inline frontend/src/Login/ForgottenPassword.jsx]: I think our users may not know what data should input.
placeholder name could be username or email.

**Review by kyya — CHANGES_REQUESTED** [review]

**Comment by HanqingZ** [comment]
Since I add a verification using the regular expression, I have to ignore the lint for this part

**Review by HONGTAO-T — APPROVED** [review]

File: frontend/src/Login/SubmitButton.jsx [hunk 5]
```diff
@@ -1,21 +1,17 @@
 import React from 'react';
 import PropTypes from 'prop-types';
 import { StyleSheet, TouchableHighlight } from 'react-native';
-import Icon from 'react-native-vector-icons/FontAwesome';
+import Icon from 'react-native-vector-icons/Entypo';
```
Comment (HONGTAO-T) [inline frontend/src/Login/SubmitButton.jsx]:
```
import { Ionicons } from '@expo/vector-icons';
```

File: frontend/src/Login/SubmitButton.jsx [hunk 6]
```diff
@@ -27,10 +23,9 @@ function SubmitButton({ onPress }) {
       onPress={onPress}
     >
       <Icon
-        name="angle-right"
+        name="chevron-right"
         color="#fff"
         size={32}
-        style={styles.icon}
       />
```
Comment (HONGTAO-T) [inline frontend/src/Login/SubmitButton.jsx]:
```
    <Ionicons
      name="ios-arrow-round-forward"
      color="#fff"
      size={32}
    />
```

File: frontend/src/Login/ForgottenPassword.jsx [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+      >
+        <Text style={styles.header}>Forget Password</Text>
+        <View style={{ width: 300 }}>
+          <InputField
+            placeholder="Username or Email"
+            value={input}
+            onChangeText={(text) => setInput(text)}
+          />
+          <View style={styles.button}>
+            <View style={{ transform: [{ rotate: '180deg' }] }}>
```
Comment (HONGTAO-T) [inline frontend/src/Login/ForgottenPassword.jsx]:
```
<View style={{ transform: [{ rotateY: '180deg' }] }}>
```

**Review by HONGTAO-T — APPROVED** [review]

**Review by GuJialu — APPROVED** [review]

File: frontend/src/Login/ForgottenPassword.jsx [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+  const USERNAME_BLACKLIST = ['admin', 'root'];
+
+  function validation(email) {
+    const emailChecker = new RegExp([
+      '^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)',
+      '|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])',
+      '|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$',
+    ].join(''));
+    return emailChecker.test(String(email).toLowerCase());
+  }
```
Comment (kyya) [inline frontend/src/Login/ForgottenPassword.jsx]: If it is not related to the rendering process, I suggest moving it out of this `ForgottenPassword` function.

File: frontend/src/Login/ForgottenPassword.jsx [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+  function handleSubmit() {
+    if (input === '') {
+      return showAlert('The input field is required.');
+    }
+    if (input.length <= 3) {
+      return showAlert('Please enter a valid username.');
+    }
+    if (!validation(input) && USERNAME_BLACKLIST.indexOf(input.toLowerCase()) !== -1) {
```
Comment (kyya) [inline frontend/src/Login/ForgottenPassword.jsx]: We can use es6 ways.
```suggestion
    if (!validation(input) && USERNAME_BLACKLIST.includes(input.toLowerCase())) {
```

**Review by kyya — APPROVED** [review]

**Review by evasimonyi — APPROVED** [review]
very nice! 👍

**Pull request merged**